cassiopeia/solver/GreedySolver.py

```diff
@@ -44,129 +48,18 @@ def solve(self) -> nx.DiGraph:
-
-    if verbose:
-        logging.info(f"{cm.shape[0] - cm_drop.shape[0]} duplicate samples removed")
-
-    return cm_drop
-
-def map_states(cm, unedit_char, missing_char) -> Tuple[pd.DataFrame, Dict[int, str]]:
-    """Maps the characters in a character matrix to the effective states.
-    """
-    state_map = {}
-    index = 1
-    unique_states = sorted(pd.unique(cm.values.ravel()))
-    if missing_char in unique_states:
-        unique_states.remove(missing_char)
-        state_map[-1] = missing_char
-        cm = cm.replace(missing_char, -1)
-    unique_states.remove(unedit_char)
-    state_map[0] = unedit_char
-    cm = cm.replace(unedit_char, 0)
-    for i in unique_states:
-        state_map[index] = i
-        cm = cm.replace(i, index)
-        index += 1
-
-    return cm, state_map
-
-def consensus_vec(vec1, vec2):
-    """Builds a consensus vector from two, obeying Camin-Sokal Parsimony.
-    """
-    assert(len(vec1) == len(vec2))
-    consensus = [0] * len(vec1)
-    for i in range(len(vec1)):
-        if vec1[i] == -1 and vec2[i] != -1:
-            consensus[i] = vec2[i]
-        if vec2[i] == -1 and vec1[i] != -1:
-            consensus[i] = vec1[i]
-        if vec1[i] == vec2[i] and vec1[i] != 0:
-            consensus[i] = vec1[i]
-    return consensus
-
-def annotate_internal_nodes(network, node, char_map):
-    """Annotates the character vectors of the internal nodes of a reconstructed
-    network from the samples, obeying Camin-Sokal Parsimony.
-    """
-    if network.out_degree(node) == 0:
-        return
-    vecs = []
-    for i in network.successors(node):
-        annotate_internal_nodes(network, i, char_map)
-        vecs.append(char_map[i])
-    assert(len(vecs) == 2)
-    consensus = consensus_vec(vecs[0], vecs[1])
-    char_map[node] = consensus
-    return
-
-def collapse_edges_recon(network, node, char_map):
-    """A helper function to help collapse edges in a tree.
-    """
-    if network.out_degree(node) == 0:
-        return
-    to_remove = []
-    to_collapse = []
-    for i in network.successors(node):
-        to_collapse.append(i)
-    for i in to_collapse:
-        collapse_edges_recon(network, i, char_map)
-        if char_map[i] == char_map[node]:
-            for j in network.successors(i):
-                network.add_edge(node, j)
-            to_remove.append(i)
-    for i in to_remove:
-        network.remove_node(i)
-    return
 
-def collapse_tree_recon(T, cm):
-    """Collapses non-informative edges (edges with 0 mutations) in a tree.
-    """
-    leaves = [n for n in T if T.out_degree(n) == 0 and T.in_degree(n) == 1]
-    char_map = {}
-    for i in leaves:
-        char_map[i] = list(cm.iloc[i,:])
-    
-    root = [n for n in T if T.in_degree(n) == 0][0]
-    annotate_internal_nodes(T, root, char_map)
-    collapse_edges_recon(T, root, char_map)
-    
\ No newline at end of file
+        _solve(range(self.prune_cm.shape[0]))
+        utils.collapse_tree_recon(self.recon, self.prune_cm)
```

Comment:
Let's have one collapse_tree function. All trees should be able to passed to `collapse_tree`.